### PR TITLE
Implement basic schedule generator

### LINKED ISSUE #80: Implement basic schedule generator

### PR DESCRIPTION
**Pull request by glasierr**
resolves #80

### CODE REVIEW

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.*;
+
+import static eu.socialedge.hermes.backend.schedule.domain.BasicScheduleGenerator.Direction.INBOUND;
+import static eu.socialedge.hermes.backend.schedule.domain.BasicScheduleGenerator.Direction.OUTBOUND;
+
+@Builder
+@Setter @Accessors(fluent = true)
+public class BasicScheduleGenerator implements ScheduleGenerator {
+
+    private @NonNull String description;
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:37]: What is `@NonNull` for here?

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    private @NonNull Duration minLayover;
+
+    @Override
+    public Schedule generate() {
+        return new Schedule(description, availability, generateTrips());
+    }
+
+    private List<Trip> generateTrips() {
+        val timePoints = generateTimePoints();
+        timePoints.sort(Comparator.comparing(TimePoint::time));
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:60]: Consider passing comparator to `generateTimePoints` method

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+        val trips = new ArrayList<Trip>();
+        for (int vehId = 1; hasNotServicedTimePoints(timePoints); vehId++) {
+            TimePoint startPoint = getNextNotServicedTimePoint(timePoints);
+            trips.addAll(generateVehicleTrips(vehId, startPoint, timePoints));
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:65]: imho it's not good to modify startPoint inside `generateVehicleTrips`

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            route,
+            vehicleId,
+            route.stations().get(route.stations().size() - 1).name(),
+            calculateStops(timePoint.time(), route, averageSpeed, dwellTime));
+    }
+
+    private List<Stop> calculateStops(LocalTime startTime, Route route, Quantity<Speed> averageSpeed, Duration dwellTime) {
+        val stops = new ArrayList<Stop>();
+
+        val speedValue = averageSpeed.to(Units.METRE_PER_SECOND).getValue().longValue();
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java]: speedValue -> averageSpeedVal :) ?

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+        for (int i = 0; i < route.stations().size(); i++) {
+            val station = route.stations().get(i);
+            val distTravelled = route.shape().shapePoints().stream()
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java]: Consider extracting to another method like `Optional<Quantity<Length>> distanceTraveled(Route r)`

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    private boolean isInTimeToTravel(LocalTime from, TimePoint toPoint) {
+        return !Duration.between(from, toPoint.time()).minus(minLayover).isNegative();
+    }
+
+    private static boolean hasNotServicedTimePoints(List<TimePoint> timePoints) {
+        return timePoints.stream().anyMatch(timePoint -> !timePoint.isServiced());
+    }
+
+    private static TimePoint getNextNotServicedTimePoint(List<TimePoint> timePoints) {
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:152]: findNextNotServicedTimePoint

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/Schedule.java [hunk 7]
```diff
@@ -78,6 +78,6 @@ public void removeTrip(Trip trip) {
     }
 
     public Collection<Trip> trips() {
```
Comment (wingsofovnia) [inline backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/Schedule.java]: Make it List<Trip> to be consistent with other classes eg https://github.com/socialedge/hermes/blob/develop-v2/backend-transit/src/main/java/eu/socialedge/hermes/backend/transit/domain/Route.java#L104

**Review by wingsofovnia — CHANGES_REQUESTED** [review]

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java (hunk 1 above)
Comment (glasierr) [reply backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:37]: It is required for Schedule

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java (hunk 2 above)
Comment (glasierr) [reply backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:60]: I'll just move this sorting to that methot

File: backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java (hunk 6 above)
Comment (glasierr) [reply backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/BasicScheduleGenerator.java:152]: I understand "find" like looking for something that might not exist. But this method assumes that next not services time point exists, so just "gets" it

**Review by wingsofovnia — APPROVED** [review]

**Pull request merged**